File: 09.19/프로그래머스/이아름/하노이의 탑.cpp
#include <string>
#include <vector>

using namespace std;

vector<vector<int>> answer;

void hanoi(int n, int start, int to, int via) {
	if (n == 1) {
		answer.push_back({ start, to });
		return;
	}
	hanoi(n - 1, start, via,to); //n-1까지를 경유지에 둠
	answer.push_back({ start,to }); //n을 목적지로 옮김
	hanoi(n - 1, via, to, start); //n-1까지를 목적지로 옮김
}

/*
	1							1
 	2     ->      1    ->		2
	3			  2 3			3
	. . .		. . .		. . .
*/

vector<vector<int>> solution(int n) {
	hanoi(n, 1, 3, 2);
	return answer;
}

File: 09.19/프로그래머스/조수연/하노이의 탑.kt
package com.ssafy.algorithm

import kotlin.collections.ArrayList

fun main() {
    val sl = Solution()
    val n = 2
    println(sl.solution(n))
}

class Solution {
    lateinit var move : ArrayList<IntArray>
    fun solution(n: Int): Array<IntArray> {
        var answer = arrayOf<IntArray>()
        move = ArrayList()
        hanoi(n,1,3,2)
        answer = Array(move.size){IntArray(2)}
        for (i in move.indices){
            answer[i] = move[i]
        }
        return answer
    }

    fun hanoi(n : Int, start : Int, end : Int, sub : Int){ // 원판 이동
        if (n == 1){
            move.add(intArrayOf(start,end))
            return
        }

        hanoi(n-1,start,sub,end) //시작 지점에서 중간지점까지 이동
        move.add(intArrayOf(start,end))
        hanoi(n-1,sub,end,start) //중간 지점에서 끝 지점까지 이동
    }
}


